### PR TITLE
Add cppfunction for lambda function

### PR DESCRIPTION
**Pull request by ZipoChan**
This pr add cppfunction for lambda function.
You can use a GenerateFunction such as:
function<packToken(TokenMap)> GenerateGFunction() 
to generate a lambda function and use CppFunction to add it into the tokenmap.

The lambda function is better than normal function pointer.

### CODE REVIEW

**Comment by ZipoChan** [comment]
@VinGarcia @MariadeAnton @wille @bamos 
Some one to review it?

**Comment by ZipoChan** [comment]
This is a example for CppFunction with lambda function.
```` C
#include <iostream>
#include"shunting-yard.h"
using namespace std;

int function1(int a, int b, int c) {
	return a + b + c;
}

function<packToken(TokenMap)> GenerateGFunction(int param3) {
	auto GFunc = [=](TokenMap Scope) {
			return (packToken)function1( Scope["A"].asInt(), Scope["B"].asInt(),param3 );
	};
	return GFunc;
}

int main()
{
    TokenMap tm = TokenMap();
	list<string> ls = { "A","B" };
	tm["Gfunc"]= CppFunction(GenerateGFunction(10), ls);
	calculator c1("Gfunc(10,4)"); //result should be 24
    cout << c1.eval(tm).asInt();
    return 0;
}

````

**Comment by VinGarcia** [comment]
I will read it this week thanks for submitting

Em dom, 3 de mai de 2020 06:59, Zipo Chen <notifications@github.com>
escreveu:

File: functions.cpp [hunk 1]
```diff
@@ -87,17 +87,24 @@ packToken Function::call(packToken _this, const Function* func,
 }
 
 /* * * * * class CppFunction * * * * */
+CppFunction::CppFunction() {
+    this->_name = "";
+    this->isStdFunc = false;
+    this->func = NULL;
+}
```
Comment (VinGarcia) [inline functions.cpp:94]: Is this constructor really any useful? If someone tries to call it it might cause a segmentation fault or I am being paranoid?

**Comment by VinGarcia** [comment]
Very cool addition and the code is very clean, thanks!

I also noticed you just found a way to create CppFunctions inline, and I think this is awesome, I wrote a sample based on your example to describe what I mean:

```c++
#include <iostream>

#include"shunting-yard.h"

int main() {
  TokenMap tm = TokenMap();

  // Inline CppFunction:
  tm["inlineFunc"] = CppFunction([=](TokenMap Scope) {
    return packToken(Scope["A"].asInt() + Scope["B"].asInt() + 10);
  }, (std::list<std::string>){ "A", "B" });

  calculator c1("inlineFunc(10, 4)");

  std::cout << c1.eval(tm) << std::endl; // result should be 24

  return 0;
}
```

> Note I am using `std::`, because this is the pattern I was using when I write the docs, so if I am going to add this to the docs I will have to keep the same idiom.

**Comment by VinGarcia** [comment]
Also, I was thinking, if we change the argument order at least for the function that accepts an `args_t` we might get an even better looking inline function, tell me what you think about it, e.g.:

```c++
#include <iostream>

#include"shunting-yard.h"

int main() {
  TokenMap tm = TokenMap();

  // Inline CppFunction:
  tm["inlineFunc"] = CppFunction({ "A", "B" }, [=](TokenMap Scope) {
    return packToken(Scope["A"].asInt() + Scope["B"].asInt() + 10);
  });

  calculator c1("inlineFunc(10, 4)");

  std::cout << c1.eval(tm) << std::endl; // result should be 24

  return 0;
}
```

I know it breaks the pattern used on the other functions, but it does look good.

**Comment by VinGarcia** [comment]
Finally @ZipoChan you said that "The lambda function is better than normal function pointer.", you mean because it is easier to use, or it is also better in performance?

**Comment by ZipoChan** [comment]
> Finally @ZipoChan you said that "The lambda function is better than normal function pointer.", you mean because it is easier to use, or it is also better in performance?

The lambda functions can capture variables,which the normal functions can't. For example,we can create a lambda function which record the variable. In this codes,we use GenerateGFunction to create a lambda function which captures the variable param3(in this case is 10),which means every GFunc store a variable param3 and the value of the param3(in this case is 10).As a constract, normal functions can't store or capture variables.

``` C
#include <iostream>
#include"shunting-yard.h"
using namespace std;

int function1(int a, int b, int c) {
	return a + b + c;
}

function<packToken(TokenMap)> GenerateGFunction(int param3) {

//The GFunc capture the variable param3.Even it can't get param3 from parameter
	auto GFunc = [=](TokenMap Scope) {
			return (packToken)function1( Scope["A"].asInt(), Scope["B"].asInt(),param3 );
	};
	return GFunc;
}

int main()
{
    TokenMap tm = TokenMap();
	list<string> ls = { "A","B" };
	tm["Gfunc"]= CppFunction(GenerateGFunction(10), ls);
	calculator c1("Gfunc(10,4)"); //result should be 24
    cout << c1.eval(tm).asInt();
    return 0;
}
```

File: functions.cpp (hunk 1 above)
Comment (ZipoChan) [reply functions.cpp:94]: In some cases,an error accurred because of the cppfunction doesn't has a default constructor.So I add it to the codes.

**Comment by ZipoChan** [comment]
In this case,I need to pass a varaible（RoleManager*） rm into the Cppfunction.We can't do it with normal functions,but we can do it with lambda funcitons.

``` C
bool BuiltinOperators::GFunction(RoleManager* rm, vector<string> ils)
{
	int lenargs = ils.size();
	string name1 = ils[0];
	string name2 = ils[1];

	if (rm == NULL) {
		return name1 == name2;
	}
	else if (lenargs == 2) {
		return rm->HasLink(name1, name2, {});
	}
	else {
		string domain = ils[2];
		rm->PrintRoles();
		return rm->HasLink(name1, name2, { domain });
	}
}


function<packToken(TokenMap)> BuiltinOperators::GenerateGFunc(RoleManager* rm) {
	function <packToken(TokenMap)> GFunc = [=](TokenMap Scope) {
		if (Scope["C"].str() != "None") {
			return (packToken)GFunction(rm, { Scope["A"].asString(),Scope["B"].asString() ,Scope["C"].asString() });
		}
		else if (Scope["B"].str() != "None") {
			return (packToken)GFunction(rm, { Scope["A"].asString(),Scope["B"].asString() });
		}
	};
	return GFunc;
}
```

**Comment by ZipoChan** [comment]
> Also, I was thinking, if we change the argument order at least for the function that accepts an `args_t` we might get an even better looking inline function, tell me what you think about it, e.g.:
> 
> ```c++
> #include <iostream>
> 
> #include"shunting-yard.h"
> 
> int main() {
>   TokenMap tm = TokenMap();
> 
>   // Inline CppFunction:
>   tm["inlineFunc"] = CppFunction({ "A", "B" }, [=](TokenMap Scope) {
>     return packToken(Scope["A"].asInt() + Scope["B"].asInt() + 10);
>   });
> 
>   calculator c1("inlineFunc(10, 4)");
> 
>   std::cout << c1.eval(tm) << std::endl; // result should be 24
> 
>   return 0;
> }
> ```
> 
> I know it breaks the pattern used on the other functions, but it does look good.

Maybe we can add a polymorphic function?Both are acceptable.

**Comment by VinGarcia** [comment]
> Maybe we can add a polymorphic function? Both are acceptable.

I agree, can you make this change when you have time?

I can do it too, but I am not sure when I will be able to do so.

**Comment by ZipoChan** [comment]
@VinGarcia 

I have added the polymorphic function.

**Comment by ZipoChan** [comment]
@VinGarcia Can you merge the PR,or should I do more changes?

**Comment by VinGarcia** [comment]
I am sorry for taking so long to answer @ZipoChan, I have little time for my open-source projects lately =/.

There are 3 things we'll have to do sometime in the future:

1. Adding tests on the file `test-shunting-yard.cpp`
2. Removing some warnings I am getting when compiling; but since this is compiler dependent you might no be seeing them so it might be better for me to fix them after the merge (if you are seeing any and you can fix them it would be nice too for other programmers with your setup).
3. Writing the docs, I'll do this when I can, but it's after the merge.

Also, I am a little concerned about the one problem I mentioned on one of the comments, I will add a little more info about why I a worried.

In either case, taking this much time to respond is inpolite from my part, so if you don't have time to finish this I can accept your PR right now and I will fix these issues eventually.

File: functions.cpp (hunk 1 above)
Comment (VinGarcia) [reply functions.cpp:94]: Hello @ZipoChan, one of the things I want for this library if for it to protect our users from unexpected and hard-to-track segmentation faults, and using this constructor might be a very easy way for the user of the library to cause one, here a sample code that could cause this problem:

```c++
#include <iostream>

#include"shunting-yard.h"

int main() {
  TokenMap tm = TokenMap();

  tm["inlineFunc"] = CppFunction();

  calculator c1("inlineFunc(10, 4)");

  std::cout << c1.eval(tm) << std::endl; // result should be 24

  return 0;
}
```

In the cases you mentioned that an error occurred I am imagining that these are static compilation errors, which are far easier to debug than segfaults. So I would prefer for the library to be a little harder to use than far harder to debug afterward.

Again I am sorry for taking so long to respond.

**Comment by VinGarcia** [comment]
@ZipoChan can I merge the PR then? Thanks for the contribution

**Pull request merged**